File: src/main/java/algorithm/Person.java
package algorithm;

import java.util.Date;

public class Person {

	private final String name;
	private final Date birthDate;

	public Person(String name, Date birthDate) {
		this.name = name;
		this.birthDate = birthDate;
	}

	Boolean isYoungerThan(Person anotherPerson) {
		return birthDate.getTime() < anotherPerson.birthDate.getTime();
	}

	long ageDifferenceWith(Person anotherPerson) {
		return birthDate.getTime() - anotherPerson.birthDate.getTime();
	}

	Person youngest(Person anotherPerson) {
		return isYoungerThan(anotherPerson) ? this : anotherPerson;
	}

	Person oldest(Person anotherPerson) {
		return isYoungerThan(anotherPerson) ? anotherPerson : this;
	}

	int nameLength() {
		return name.length();
	}
}

File: src/main/java/algorithm/Person.kt
package algorithm

import java.util.Date

class Person(private val name: String, private val birthDate: Date) {

    fun ageDifferenceWith(anotherPerson: Person) = birthDate.time - anotherPerson.birthDate.time

    fun youngest(anotherPerson: Person) = if (isYoungerThan(anotherPerson)) this else anotherPerson

    fun oldest(anotherPerson: Person) = if (isYoungerThan(anotherPerson)) anotherPerson else this

    internal fun isYoungerThan(anotherPerson: Person) = birthDate.time < anotherPerson.birthDate.time

    fun nameLength() = name.length
}
